File: Engine/Android/core/src/test/kotlin/despia/engine/SnackbarConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The SNACKBAR conformance runner — executes OpenSource/Conformance/overlays/snackbar.json
 * through this runtime (the TS and Swift runners run the SAME file). The law and the reasoning
 * are in that corpus's README.
 *
 * The rule worth restating: `show` resolves on OUTCOME, so every card must settle EXACTLY ONCE.
 * A card that never settles is a leaked promise; one that settles twice is a double undo.
 *
 * Missing corpus = loud failure; a silently-skipped conformance suite is how drift starts.
 */
class SnackbarConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/overlays/snackbar.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/overlays/snackbar.json not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(name: String): List<Map<String, Any?>> {
        val doc = json(corpusFile().readText()).foundationValue as? Map<String, Any?>
            ?: error("snackbar.json: not a JSON object")
        return (doc[name] as? List<*> ?: error("snackbar.json: no $name"))
            .map { it as Map<String, Any?> }
    }

    private fun int(any: Any?): Int = when (any) {
        is Number -> any.toInt()
        else -> error("expected a number, got $any")
    }

    @Test
    fun durationCorpus() {
        val cases = section("durationCases")
        assertTrue(cases.isNotEmpty(), "durationCases is empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "(unnamed)"
            val got = SnackbarQueue.resolveDuration(case["duration"], case["hasAction"] as? Boolean ?: false)
            assertEquals(int(case["expect"]), got, name)
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun queueCorpus() {
        val cases = section("queueCases")
        assertTrue(cases.isNotEmpty(), "queueCases is empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "(unnamed)"
            var state = SnackbarState()
            for (rawOp in case["ops"] as List<Map<String, Any?>>) {
                state = SnackbarQueue.applyOp(state, opFrom(rawOp))
            }
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["current"] as? String, state.current?.id, "$name — current")
            assertEquals(
                (expect["pending"] as List<*>).map { it as String },
                state.queue.map { it.id },
                "$name — pending",
            )
            assertEquals(
                (expect["settled"] as List<*>).map { it as Map<String, Any?> }
                    .map { (it["id"] as String) to (it["result"] as String) },
                state.settled.map { it.id to it.result.word },
                "$name — settled",
            )
            assertEquals(state.queue.size, state.pending, "$name — pending count")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun opFrom(raw: Map<String, Any?>): SnackbarOp {
        val id = raw["id"] as? String
        return when (raw["op"] as? String) {
            "show" -> {
                val action = (raw["action"] as? Map<String, Any?>)?.let {
                    SnackbarAction(it["label"] as? String ?: "", it["id"] as? String ?: "")
                }
                // The corpus omits `message` where it does not matter; the reducer treats an empty
                // message as a no-op, so the runner supplies the id as the text unless a case is
                // deliberately testing emptiness.
                SnackbarOp.Show(
                    SnackbarRequest(
                        id = id ?: "",
                        message = raw["message"] as? String ?: (id ?: "x"),
                        action = action,
                        duration = raw["duration"],
                        tone = raw["tone"] as? String,
                        icon = raw["icon"] as? String,
                        position = raw["position"] as? String,
                        dismissible = raw["dismissible"] as? Boolean,
                        replace = raw["replace"] as? Boolean ?: false,
                    ),
                )
            }
            "elapse" -> SnackbarOp.Elapse
            "action" -> SnackbarOp.Action
            "dismiss" -> SnackbarOp.Dismiss
            "hide" -> SnackbarOp.Hide(id)
            else -> error("unknown op ${raw["op"]}")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun liftCorpus() {
        val cases = section("liftCases")
        assertTrue(cases.isNotEmpty(), "liftCases is empty")
        for (case in cases) {
            val name = case["name"] as? String ?: "(unnamed)"
            val lift = case["lift"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val got = SnackbarQueue.resolveLift(
                SnackbarQueue.Chrome(
                    position = lift["position"] as? String,
                    safeAreaTop = int(lift["safeAreaTop"]),
                    safeAreaBottom = int(lift["safeAreaBottom"]),
                    bottomBar = int(lift["bottomBar"]),
                    fab = int(lift["fab"]),
                    keyboard = int(lift["keyboard"]),
                ),
            )
            assertEquals(expect["edge"] as String, got.edge.word, "$name — edge")
            assertEquals(int(expect["inset"]), got.inset, "$name — inset")
            assertEquals(expect["clearsHomeIndicator"] as Boolean, got.clearsHomeIndicator, "$name — clearsHomeIndicator")
        }
    }

    @Test
    fun durationParsingIsTotal() {
        for (junk in listOf(null, "", "   ", "soon", "NaN", true, emptyList<String>())) {
            assertEquals(SnackbarQueue.DEFAULT_MS, SnackbarQueue.resolveDuration(junk), "junk: $junk")
            assertEquals(SnackbarQueue.SHORT_MS, SnackbarQueue.resolveDuration(junk, true), "junk: $junk")
        }
        for (bad in listOf(Double.NaN, Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY)) {
            assertEquals(SnackbarQueue.DEFAULT_MS, SnackbarQueue.resolveDuration(bad), "bad: $bad")
        }
        assertEquals(SnackbarQueue.MIN_MS, SnackbarQueue.resolveDuration(-1_000_000))
    }

    @Test
    fun theCardNeverCoversTheHomeIndicator() {
        val values = listOf(-10_000, -1, 0, 12, 34, 400, 10_000)
        for (safeBottom in values) {
            for (keyboard in values) {
                for (bar in listOf(0, 49, -49)) {
                    val got = SnackbarQueue.resolveLift(
                        SnackbarQueue.Chrome(
                            position = "bottom", safeAreaTop = 47, safeAreaBottom = safeBottom,
                            bottomBar = bar, fab = 0, keyboard = keyboard,
                        ),
                    )
                    val safe = if (safeBottom > 0) safeBottom else 0
                    assertTrue(got.inset >= safe + SnackbarQueue.GAP, "inset ${got.inset} sits on the home indicator")
                    assertTrue(got.clearsHomeIndicator, "clearsHomeIndicator went false")
                }
            }
        }
    }

    @Test
    fun oneAtATimeAndTheQueueDrainsInOrder() {
        var state = SnackbarState()
        val ids = listOf("a", "b", "c", "d", "e")
        for (id in ids) state = SnackbarQueue.applyOp(state, SnackbarOp.Show(SnackbarRequest(id = id, message = id)))
        assertEquals("a", state.current?.id)
        assertEquals(4, state.pending)
        val seen = mutableListOf<String>()
        repeat(ids.size) {
            seen.add(state.current?.id ?: "")
            state = SnackbarQueue.applyOp(state, SnackbarOp.Elapse)
        }
        assertEquals(ids, seen)
        assertNull(state.current)
        assertEquals(ids.map { SnackbarResult.TIMEOUT }, state.settled.map { it.result })
    }

    @Test
    fun everyCardSettlesExactlyOnce() {
        var state = SnackbarState()
        state = SnackbarQueue.applyOp(
            state,
            SnackbarOp.Show(SnackbarRequest(id = "a", message = "a", action = SnackbarAction("Undo", "u"))),
        )
        state = SnackbarQueue.applyOp(state, SnackbarOp.Show(SnackbarRequest(id = "b", message = "b")))
        for (op in listOf(SnackbarOp.Action, SnackbarOp.Action, SnackbarOp.Dismiss, SnackbarOp.Elapse, SnackbarOp.Hide())) {
            state = SnackbarQueue.applyOp(state, op)
        }
        assertEquals(mapOf("a" to 1, "b" to 1), state.settled.groupingBy { it.id }.eachCount())
    }

    @Test
    fun anActionButtonWithNoLabelIsNotAnActionButton() {
        val entry = SnackbarQueue.normalize(
            SnackbarRequest(id = "a", message = "m", action = SnackbarAction("", "u")),
        )
        assertNull(entry.action)
        assertEquals(SnackbarQueue.DEFAULT_MS, entry.durationMs, "no action means the wordless default stays 2s")
    }
}

File: Web/packages/kernel/test/snackbar-conformance.test.ts
//
//  snackbar-conformance.test.ts — the snackbar corpus runner (TS lane).
//  Executes OpenSource/Conformance/overlays/snackbar.json against src/snackbar.ts; the
//  Kotlin (:core SnackbarConformanceTest) and Swift (SnackbarConformance) twins run the
//  SAME file.
//
//  Missing corpus = loud failure — a silently-skipped conformance suite is how drift starts.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, resolve, dirname } from "node:path";

import {
  resolveDuration,
  resolveLift,
  resolveEdge,
  snackbarApply,
  snackbarInitial,
  snackbarPending,
  normalizeRequest,
  SNACKBAR_GAP,
  SNACKBAR_MIN_MS,
  type SnackbarOp,
  type SnackbarState,
} from "../src/snackbar.ts";

function corpusFile(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/overlays/snackbar.json");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/overlays/snackbar.json not found");
    dir = parent;
  }
}

interface DurationCase {
  name: string;
  duration: string | number | null;
  hasAction?: boolean;
  expect: number;
}

interface QueueCase {
  name: string;
  ops: Array<Record<string, unknown>>;
  expect: { current: string | null; pending: string[]; settled: Array<{ id: string; result: string }> };
}

interface LiftCase {
  name: string;
  lift: {
    position: string;
    safeAreaTop: number;
    safeAreaBottom: number;
    bottomBar: number;
    fab: number;
    keyboard: number;
  };
  expect: { edge: string; inset: number; clearsHomeIndicator: boolean };
}

const corpus = JSON.parse(readFileSync(corpusFile(), "utf8")) as {
  durationCases: DurationCase[];
  queueCases: QueueCase[];
  liftCases: LiftCase[];
};

test("snackbar duration corpus", () => {
  assert.ok(corpus.durationCases.length > 0, "durationCases is empty");
  for (const c of corpus.durationCases) {
    assert.equal(resolveDuration(c.duration, c.hasAction === true), c.expect, c.name);
  }
});

test("snackbar queue corpus", () => {
  assert.ok(corpus.queueCases.length > 0, "queueCases is empty");
  for (const c of corpus.queueCases) {
    let state: SnackbarState = snackbarInitial();
    for (const raw of c.ops) {
      // The corpus omits `message` where it does not matter; the reducer treats an empty
      // message as a no-op, so the runner supplies the id as the text unless a case is
      // deliberately testing emptiness.
      const op = { message: raw.message === undefined ? String(raw.id ?? "x") : raw.message, ...raw } as SnackbarOp;
      state = snackbarApply(state, op);
    }
    assert.equal(state.current === null ? null : state.current.id, c.expect.current, `${c.name} — current`);
    assert.deepEqual(state.queue.map((e) => e.id), c.expect.pending, `${c.name} — pending`);
    assert.deepEqual(state.settled, c.expect.settled, `${c.name} — settled`);
    assert.equal(snackbarPending(state), c.expect.pending.length, `${c.name} — pending count`);
  }
});

test("snackbar lift corpus", () => {
  assert.ok(corpus.liftCases.length > 0, "liftCases is empty");
  for (const c of corpus.liftCases) {
    assert.deepEqual(resolveLift(c.lift), c.expect, c.name);
  }
});

test("duration parsing is total — no input throws, everything unknown is the default", () => {
  for (const junk of [undefined, null, "", "   ", "soon", "NaN", {}, [], true]) {
    assert.equal(resolveDuration(junk as never), 2000, String(junk));
    assert.equal(resolveDuration(junk as never, true), 4000, String(junk));
  }
  for (const bad of [Number.NaN, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY]) {
    assert.equal(resolveDuration(bad), 2000, String(bad));
  }
  assert.equal(resolveDuration(-1_000_000), SNACKBAR_MIN_MS);
});

test("the edge word is total and defaults to bottom", () => {
  assert.equal(resolveEdge("top"), "top");
  assert.equal(resolveEdge(" TOP "), "top");
  for (const junk of [null, undefined, "", "bottom", "middle", "TOPMOST"]) {
    if (junk === "top") continue;
    assert.equal(resolveEdge(junk as string), junk === "top" ? "top" : "bottom", String(junk));
  }
});

test("the card never covers the home indicator, whatever geometry arrives", () => {
  const values = [-10_000, -1, 0, 12, 34, 400, 10_000, Number.NaN, Number.POSITIVE_INFINITY];
  for (const safeAreaBottom of values) {
    for (const keyboard of values) {
      for (const bottomBar of [0, 49, -49]) {
        const lift = resolveLift({
          position: "bottom",
          safeAreaTop: 47,
          safeAreaBottom,
          bottomBar,
          fab: 0,
          keyboard,
        });
        const safe = Number.isFinite(safeAreaBottom) && safeAreaBottom > 0 ? safeAreaBottom : 0;
        assert.ok(lift.inset >= safe + SNACKBAR_GAP, `inset ${lift.inset} sits on the home indicator`);
        assert.ok(lift.clearsHomeIndicator, "clearsHomeIndicator went false");
      }
    }
  }
});

test("one at a time: N shows leave N-1 waiting and drain in order", () => {
  let state = snackbarInitial();
  const ids = ["a", "b", "c", "d", "e"];
  for (const id of ids) state = snackbarApply(state, { op: "show", id, message: id });
  assert.equal(state.current?.id, "a");
  assert.equal(snackbarPending(state), 4);
  const seen: string[] = [];
  for (let i = 0; i < ids.length; i += 1) {
    seen.push(state.current?.id ?? "");
    state = snackbarApply(state, { op: "elapse" });
  }
  assert.deepEqual(seen, ids);
  assert.equal(state.current, null);
  assert.deepEqual(state.settled.map((s) => s.result), ids.map(() => "timeout"));
});

test("every card settles exactly once, whatever order the endings arrive in", () => {
  let state = snackbarInitial();
  state = snackbarApply(state, { op: "show", id: "a", message: "a", action: { label: "Undo", id: "u" } });
  state = snackbarApply(state, { op: "show", id: "b", message: "b" });
  for (const op of ["action", "action", "dismiss", "elapse", "hide"] as const) {
    state = snackbarApply(state, { op } as SnackbarOp);
  }
  const counts = new Map<string, number>();
  for (const s of state.settled) counts.set(s.id, (counts.get(s.id) ?? 0) + 1);
  assert.deepEqual([...counts.entries()].sort(), [["a", 1], ["b", 1]]);
});

test("an action button with no label is not an action button", () => {
  const entry = normalizeRequest({ id: "a", message: "m", action: { label: "", id: "u" } });
  assert.equal(entry.action, null);
  assert.equal(entry.durationMs, 2000, "no action means the wordless default stays 2s");
});
